trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/YamlCorrectnessWitnessExtractor.java

```diff
@@ -206,6 +222,8 @@ protected int genericVisit(final IASTNode node) {
 	private static final class YamlExtractedCorrectnessWitness implements IExtractedCorrectnessWitness {
 		private final HashRelation<IASTNode, ExtractedWitnessInvariant> mInvariants = new HashRelation<>();
 		private final HashRelation<IASTNode, ExtractedFunctionContract> mFunctionContracts = new HashRelation<>();
+		private final HashRelation<IASTNode, ExtractedGhostUpdate> mGhostUpdates = new HashRelation<>();
```

Comment:
This seems like it does not preserve the order of ghost updates at a given location. But that could be important, right? Maybe a `Map<IASTNode, List<ExtractedGhostUpdate>>` would be better here.

Reply:
Wasn't this our idea that the ghost updates don't have a particular order?

Reply:
But one ghost update may use the value of another ghost  variable assigned at the same location. So what would that mean exactly in terms of the semantics of witnesses? 

- One possibility: _The witness must be valid for all possible orders of ghost updates._

  This is still a bit unclear (e.g. may the order change if the ghost updates are executed multiple times?), and seems unnecessarily complicated.

- Another possibility: _For validity of witnesses, we require that all ghost updates at a certain location commute._

  Or more syntactically: _We require that no ghost variable is assigned twice at the same location, and no ghost variable that is assigned at this location is used in another ghost assignment at the same location._

  Then we should actually perform a check for such a condition.

I would prefer fixing an order of ghost updates to both possibilities above, as it seems the most intuitive and least restrictive solution. The natural order to choose would be the order of the entries in the witness.

---

Independent of this question (regarding witness semantics and validation) is the fact that when we generate a witness, for all algorithms we are _currently_ considering, the order happens not to matter.